### PR TITLE
WooCommerce: Add api plan for product edits.

### PR DESCRIPTION
**Pull request by coderkevin**
This adds the beginning of the code to convert the product edits into
sequential API calls.

There's a lot to be added from here, but the basic idea is to keep a sequential list of operations that need to happen in order to process our edit state onto the server.

Reviewers: Please take a look over the structure of this and if you have questions feel free to ask @justinshreve or myself about it.

An example of what an API plan would look like for a product that creates 3 new categories and 4 variations along with itself would look like this:
```
apiPlan: [
	{
		name: "Creating X Product Category",
		operation: { name: "createProductCategory", siteId: 1337, id: catIdX },
		startTime: XXXXXX,
		endTime: XXXXX,
		result: 'success',
	},
	{
		name: "Creating Y Product Category",
		operation: { name: "createProductCategory", siteId: 1337, id: catIdY },
		startTime: XXXXXX,
		endTime: XXXXX,
		result: 'success',
	},
	{
		name: "Creating Z Product Category",
		operation: { name: "createProductCategory", siteId: 1337, id: catIdZ },
		startTime: XXXXXX,
		endTime: XXXXX,
		result: 'success',
	},
	{
		name: "Creating Product",
		operation: { name: "createProduct", siteId: 1337, id: prodId1 }
		startTime: XXXXXX,
		endTime: XXXXX,
		result: 'success',
	},
	{
		name: "Creating Red/Small Variation",
		operation: { name: "createVariation", siteId: 1337, id: variationIdRS }
		startTime: XXXXXX,
	}
	{
		name: "Creating Red/Large Variation",
		operation: { name: "createVariation", siteId: 1337, id: variationIdRL }
	}
	{
		name: "Creating Blue/Small Variation",
		operation: { name: "createVariation", siteId: 1337, id: variationIdBS }
	}
	{
		name: "Creating Blue/Large Variation",
		operation: { name: "createVariation", siteId: 1337, id: variationIdBL }
	}
]
```

To Test:

1. Run `make test` and ensure all tests pass.

### CODE REVIEW

**Comment by DanReyLop** [comment]
An important limitation of this "API plan" is that it only allows for sequential requests. Most of these requests can be parallelized, and specially for the variations that would be a significant gain. I think it's worth it to try to architect this so it supports a more complex hierarchy of operations (I would probably be able to steal this code for the Shipping Zones stuff too).

**Comment by coderkevin** [comment]
Hi Dan,

Yeah, we definitely want to make this parallel, but it was getting really complex in our heads so we decided to go serial for now. But a step in the plan could be an array of steps that could be carried out in parallel. Good thinking!

Also, I figure we will reuse a lot of this for a lot of the other transactional code we have to write for shipping, orders, coupons, etc.

**Comment by belcherj** [comment]
@coderkevin I like this as is with a todo to:

> But a step in the plan could be an array of steps that could be carried out in parallel.

Should we code review as is or what for that?

File: client/extensions/woocommerce/state/ui/products/api-plan/actions.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+*/
+
+export function clearApiPlan() {
+	return {
+		type: WOOCOMMERCE_EDIT_PRODUCT_APIPLAN_CLEAR,
+	};
+}
+
+// The functions below are used internally and for testing.
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/api-plan/actions.js]: How about adding a jsdoc block and `@private` on these?

Comment (coderkevin) [inline client/extensions/woocommerce/state/ui/products/api-plan/actions.js]: I just tried this and if I add a jsdoc block (even with @private) it wants params and returns on it. These things are very likely to change, so I'd like to just leave this off for now.

**Comment by coderkevin** [comment]
> subsequent PR include a means of providing callbacks to be hit on 1) successful plan completion and 2) on plan step failure

I've been thinking about this and I'm not sure if it's best to do callbacks for this stuff, or maybe we should be updating the redux state for renderers and such to take advantage of.

> Please also consider (i.e. in a future PR) whether successful plan completion should dispatch a plan clear, and whether attempting to create a plan when a unfinished plan exists in state should result in an exception.

Good thoughts, although this may be plan-specific. There may be scenarios where more than one plan at a time is valid.

Seems like we have much more to discuss about this

**Pull request merged**